lib/spack/spack/database.py

```diff
@@ -766,45 +828,64 @@ def activated_extensions_for(self, extendee_spec, extensions_layout=None):
                 continue
             # TODO: conditional way to do this instead of catching exceptions
 
-    def query(self, query_spec=any, known=any, installed=True, explicit=any):
-        """Run a query on the database.
+    def query(self, query_spec=any, **kwargs):
```

Comment:
same question, why not leave the arguments explicitly? Yes it makes the signature bulkier, but it's much easier to understand the function if you don't have to constantly ask yourself what else is contained ink kwargs

Reply:
Because it forces users, in both python 2 and 3, to pass optional arguments by keyword (which makes it more readable at call site). If you look at this:
```python
db.query(spec, True, False, False)
```
are you sure it's easier to understand than:
```python
db.query(spec, known=True, installed=False, explicit=False)
```
? The longer signature is a another reason, but in my opinion of secondary importance with respect to the above.

Reply:
Yeah I agree with you that the second one is easier to read, I was thinking about writing another `db.query` statement, but then the kwargs are documented, so it is fine, I'll probably find what I need.

Reply:
I actually agree with @healther -- let's not change this everywhere just to force users to do this.  I don't think most calls to these functions use positional arguments anyway, and we can either add a style check for this, or catch it in review.